### DISCUSSION TITLE
미션을 진행하면서 랜덤 값을 테스트한 경험을 알려주세요.

### DISCUSSION THREAD

**Discussion by 3Juhwan** [2024-woowacourse-study/level-interview#110] [Clean Code]
.

**Comment by jimi567** [comment]
(경험이니깐, 구어체)
사다리 타기 도메인을 구현할 때, 사다리의 다리가 생성의 요구 사항이 랜덤성에 의존적인 내용이라, 이를 테스트 하기 위해 전략 패턴을 사용한 경험이 있습니다.
전략 패턴을 통해, 실제 프로덕션 코드에서는 요구 사항처럼 랜덤성에 의존적인 로직을 주입하여, 구현하였고, 테스트단에서는 제가 예측할 수 있는 값을 주입하여, 해당 도메인 로직이 정상 작동하는 지를 테스트한 경험이 있습니다.
추후에, 테스트만을 위해 전략 패턴으로 해당 로직을 구현하다 보니, 프로덕션에 코드의 복잡도가 테스트로 인해 높아진다는 것을 인지하고, 사다리 클래스의  
테스트 생성자를 오버로딩하여, 테스트로 인한 프로덕션 코드의 영향을 최소화 하도록 리팩터링을 진행하기도 했습니다.

**Comment by nayonsoso** [comment]
자동차 미션을 할 때였습니다.
자동차 미션의 요구 사항은 0~9 사이의 랜덤한 정수를 만들고, 정수가 4 이상일 때만 움직이게 하는 것입니다.
이를 위해서 3이면 이동하지 않고, 4이면 이동하는 기능을 테스트 해야 했습니다.

처음 작성한 코드는 자동차 객체 안에서 바로 랜덤 값을 만드는 것이었습니다.
하지만 랜덤한 값은 테스트할 수는 없다고 생각이 들었습니다.
그래서 저는 테스트가 어려운 '난수값 생성' 부분을 자동차 객체로부터 분리하고, 
다른 것으로 대체될 수 있게 함수형 인터페이스로 만들어 주었습니다.
그리고 자동차 객체가 함수형 인터페이스를 외부로부터 주입받을 수 있게 했습니다.

테스트 코드에서는 람다 함수를 함수형 인터페이스의 자리에 넣어주었고,
이를 통해 자동차의 움직임을 테스트할 수 있었습니다.

**Comment by GIVEN53** [comment]
자동차 이동 -> 랜덤으로 생성한 정수에 의존
테스트 어려움

랜덤 생성을 전략 패턴으로 구현
테스트 쉬워짐
전략이 바뀌면 구현체만 갈아끼우면 됨
